### DISCUSSION TITLE
Does uv_write2() free the client handle memory?

### DISCUSSION THREAD

**Discussion by Qix-** [libuv/libuv#3577] [Q&A] [answered]
Looking at `uv_write2()` for transferring handles between `uv_loop_t` (in my case, across threads), the example echo server doesn't seem the free or close the client instance.

https://github.com/libuv/libuv/blob/47e0c5c575e92a25e0da10fc25b2732942c929f3/docs/code/multi-echo-server/main.c#L43

Note that the callback is passed `NULL` here, which means that unless `uv_write2()` frees the memory associated with it, the memory leaks in the demo.

Is that the case? Or does `uv_write2()` own the pointer of the handle?

**Comment by bnoordhuis** [comment, marked as answer]
uv_write2() does not take ownership. `dummy_buf` points to static read-only memory so there is nothing to free.

**Reply by Qix-** [reply]
Wasn't referring to `dummy_buf` - the `client` is malloc'd on 37.

https://github.com/libuv/libuv/blob/47e0c5c575e92a25e0da10fc25b2732942c929f3/docs/code/multi-echo-server/main.c#L37

**Reply by bnoordhuis** [reply]
Oh yes, that's a memory leak.

(The guide is an external, unvetted contribution. PR welcome.)

**Reply by Qix-** [reply]
Okay so once the handle is sent via `uv_write2()` it can be freed after the callback fires, right? Does it need to be `uv_close()`'s or just the memory needs to be freed? Just making sure.

Thanks again for the info :)

**Reply by bnoordhuis** [reply]
It needs to be uv_close'd. Only when the uv_close callback is called, is it safe to free() the memory.

**Reply by Qix-** [reply]
Thanks!
